### ISSUE TITLE
noPropertyAccessFromIndexSignature should not throw errors when property access is a normal identifier (TS4111)

### ISSUE THREAD

**Issue by NullVoxPopuli** [microsoft/TypeScript#55243]
# Bug Report

### 🔎 Search Terms

- noPropertyAccessFromIndexSignature

### 🕗 Version & Regression Information

- version: 5.1.6

problem noticed when `noPropertyAccessFromIndexSignature` started getting used across codebases

Playgrounds: 
[5.1.6](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&noPropertyAccessFromIndexSignature=true#code/C4TwDgpgBAIghsOUC8UBKEDGB7ATgEwB4BnYXASwDsBzAGilIpoD4BuAKE4BsJgoAPAFywESVAG8Avh3b8A2gHIAZtmxQARnFwa4ALwUBdFFAUBWBTP4A6FWtQKALBc7XbrIA)
[nightly](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&noPropertyAccessFromIndexSignature=true&ts=5.2.0-dev.20230802#code/C4TwDgpgBAIghsOUC8UBKEDGB7ATgEwB4BnYXASwDsBzAGilIpoD4BuAKE4BsJgoAPAFywESVAG8Avh3b8A2gHIAZtmxQARnFwa4ALwUBdFFAUBWBTP4A6FWtQKALBc7XbrIA)

Please keep and fill in the line that best applies:
-->
- ~This is a crash~
- ~This changed between versions ______ and _______~
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about ____everything_____
- ~I was unable to test this on prior versions because _______~

### 💻 Code

```ts
type Data = Record<string, string>;

let x: Data = {};

x['foo bar baz'] = '5';

x.foo = '4';
// ^ error 
//   Property 'foo' comes from an index signature, so it must be accessed with ['foo'].(4111)

x.foo;
// ^ error 
//   Property 'foo' comes from an index signature, so it must be accessed with ['foo'].(4111)
```

### 🙁 Actual behavior

the `noPropertyAccessFromIndexSignature` setting is too aggressive and isn't helpful for objects which use identifier keys

### 🙂 Expected behavior

I almost want there to be a way to specify that keys can be valid identifiers. It seems that `noPropertyAccessFromIndexSignature` is designed to protect against/for/adhere-to-correctness the fact that "string" (as used in the Key in Record, above) means, literally, "any string", which could include non-identifiers, such as `foo bar baz`, or `foo.bar` -- these can be valid keys on an object, but we don't have a way to express that we can manage different kinds of keys.

Sure, an object, by default, is "anything goes" with its keys, but the type system has no way to narrow a key to an identifier.

If I could somehow say: 
```ts
type Data = Record<Identifier, string>;
```
then my problem would be solved, and my example could would provide (non) errors like so:

```ts
let x: Data = {};

x['foo bar baz'] = '5';
// ^ error
//   Property 'foo bar baz' does not match type Identifier. ( error code )

x.foo = '4';
// no error

x.foo;
// no error
```

**Comment by RyanCavanaugh** [comment]
> the noPropertyAccessFromIndexSignature setting is too aggressive and isn't helpful for objects which use identifier keys

I'm really confused about what you're asking for. It's already the case that the only time you can write `x.foo` is when `foo` is a legal identifier. If you don't like the behavior, turn off `noPropertyAccessFromIndexSignature`, this is literally only behavior it controls.

It seems like you're asking for a completely different feature entirely, not describing a defect in NPAFIS

**Comment by NullVoxPopuli** [comment]
that is correct, yes

**Comment by RyanCavanaugh** [comment]
OK, please use the "Feature request" template on a new issue. Thanks!
